packages/vue-urql/src/utils.ts

```diff
@@ -1,45 +1,142 @@
-import type { GraphQLRequest, AnyVariables } from '@urql/core';
-import type { Ref, ShallowRef } from 'vue';
-import { isRef } from 'vue';
+import type {
+  AnyVariables,
+  Client,
+  CombinedError,
+  DocumentInput,
+  Operation,
+  OperationContext,
+  OperationResult,
+  OperationResultSource,
+} from '@urql/core';
+import { createRequest } from '@urql/core';
+import type { Ref } from 'vue';
+import {
+  watchEffect,
+  isReadonly,
+  computed,
+  readonly,
+  ref,
+  shallowRef,
+  isRef,
+} from 'vue';
+import type { UseSubscriptionArgs } from './useSubscription';
+import type { UseQueryArgs } from './useQuery';
 
 export type MaybeRef<T> = T | (() => T) | Ref<T>;
 export type MaybeRefObj<T> = T extends {}
   ? { [K in keyof T]: MaybeRef<T[K]> }
   : T;
 
-export const unref = <T>(maybeRef: MaybeRef<T>): T =>
+const unwrap = <T>(maybeRef: MaybeRef<T>): T =>
   typeof maybeRef === 'function'
     ? (maybeRef as () => T)()
     : maybeRef != null && isRef(maybeRef)
     ? maybeRef.value
     : maybeRef;
 
-export interface RequestState<
-  Data = any,
-  Variables extends AnyVariables = AnyVariables,
-> {
-  request: GraphQLRequest<Data, Variables>;
-  isPaused: boolean;
-}
-
-export function createRequestState<
-  Data = any,
-  Variables extends AnyVariables = AnyVariables,
+export const createRequestWithArgs = <
+  T = any,
+  V extends AnyVariables = AnyVariables,
 >(
-  request: GraphQLRequest<Data, Variables>,
-  isPaused: boolean
-): RequestState<Data, Variables> {
-  return { request, isPaused };
-}
-
-export const updateShallowRef = <T extends Record<string, any>>(
-  ref: ShallowRef<T>,
-  next: T
+  args:
+    | UseQueryArgs<T, V>
+    | UseSubscriptionArgs<T, V>
+    | { query: MaybeRef<DocumentInput<T, V>>; variables: V }
 ) => {
-  for (const key in next) {
-    if (ref.value[key] !== next[key]) {
-      ref.value = next;
-      return;
+  let vars = unwrap(args.variables);
+  // unwrap possible nested reactive variables with `readonly()`
+  for (const prop in vars) {
+    if (Object.hasOwn(vars, prop)) {
+      const value = vars[prop];
+      if (value && (typeof value === 'object' || isRef(value))) {
+        vars = readonly(vars);
```

Comment:
Is it really necessary to wrap variables in readonly every time? I think we could avoid the additional overhead.

Reply:
@negezor
Not every time, but only in case it contains some ref deeply. For this case it will be "unwrapped" with `readonly()` and it all will be "listen" inside computed var.

```
{
  variables: {
     somevar: ref()
  }
}
```

Reply:
In comparing to original implementation (where all arguments were passed to `reactive()`), unwrapping deeply nested variables with `readonly()` shouldn't create a lot of additional overhead, as complex variables object is already overhead ;)

Reply:
@negezor that piece of code has been rewritten to not use `readonly()` but simply iterate over the object and unwrap possible refs